sw/device/lib/dif/dif_kmac.h

```diff
@@ -0,0 +1,788 @@
+  kDifKmacConfigureBadArg = kDifKmacBadArg,
+  kDifKmacConfigureError = kDifKmacError,
+
+  /**
+   * The operation failed because writes to the configuration register are
+   * disabled. This means that an operation is already in progress.
+   */
+  kDifKmacConfigureLocked,
+} dif_kmac_configure_result_t;
+
+/**
+ * Configures KMAC with runtime information.
+ *
+ * @param kmac A KMAC handle.
+ * @param config Runtime configuration parameters.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT dif_kmac_configure_result_t
+dif_kmac_configure(dif_kmac_t *kmac, dif_kmac_config_t config);
+
+/**
+ * The result of a string encoding operation.
+ */
+typedef enum dif_kmac_string_encode_result {
+  kDifKmacStringEncodeOk = kDifKmacOk,
+  kDifKmacStringEncodeBadArg = kDifKmacBadArg,
+  kDifKmacStringEncodeError = kDifKmacError,
+
+  /**
+   * Encoding failed because the provided string exceeded the maximum
+   * supported length.
+   */
+  kDifKmacStringEncodeTooLong,
+} dif_kmac_string_encode_result_t;
+
+/**
+ * Encode a customization string (S).
+ *
+ * The length of the string must not exceed `kDifKmacMaxCustomizationStringLen`.
+ *
+ * Note that this function will encode `len` bytes from `data` regardless of
+ * whether `data` is null-terminated or not.
+ *
+ * @param data String to encode.
+ * @param len Length of string to encode.
+ * @param[out] out Encoded customization string.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_string_encode_result_t dif_kmac_customization_string_init(
+    const char *data, size_t len, dif_kmac_customization_string_t *out);
+
+/**
+ * Encode a function name (N).
+ *
+ * The length of the string must not exceed `kDifKmacMaxFunctionNameLen`.
+ *
+ * Note that this function will encode `len` bytes from `data` regardless of
+ * whether `data` is null-terminated or not.
+ *
+ * @param data String to encode.
+ * @param len Length of string to encode.
+ * @param[out] out Encoded function name.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_string_encode_result_t dif_kmac_function_name_init(
+    const char *data, size_t len, dif_kmac_function_name_t *out);
+
+/**
+ * Start a SHA-3 operation.
+ *
+ * @param kmac A KMAC handle.
+ * @param mode The SHA-3 mode of operation.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_configure_result_t dif_kmac_mode_sha3_start(dif_kmac_t *kmac,
+                                                     dif_kmac_mode_sha3_t mode);
+
+/**
+ * Start a SHAKE operation.
+ *
+ * @param kmac A KMAC handle.
+ * @param mode The mode of operation.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_configure_result_t dif_kmac_mode_shake_start(
+    dif_kmac_t *kmac, dif_kmac_mode_shake_t mode);
+
+/**
+ * Start a cSHAKE operation.
+ *
+ * @param kmac A KMAC handle.
+ * @param mode The mode of operation.
+ * @param S Customization string (optional).
```

Comment:
Nit: instead of "optional", maybe be explicit with (can be null)?

Reply:
I've left it as "optional" for now since that seems to be the convention used in the other DIFs.